File: backend/src/mhglauncher/providers/device.py
from __future__ import annotations

import json
import os
import secrets
import time
import uuid
from pathlib import Path
from typing import Any

import httpx

from mhglauncher.errors import AppError

DEVICE_FP_URL = "https://public-data-api.mihoyo.com/device-fp/api/getFp"
PROFILE_VERSION = "android-v1"


class DeviceIdentity:
    def __init__(self, path: Path) -> None:
        self.path = path
        saved = self._load()
        self.device_id = str(saved.get("device_id") or uuid.uuid4())
        self.fp_device_id = str(saved.get("fp_device_id") or secrets.token_hex(8))
        self.device_name = str(saved.get("device_name") or self._random_text(12))
        self.product_name = str(saved.get("product_name") or self._random_text(6))
        self.device_fp = (
            str(saved.get("device_fp") or "")
            if saved.get("profile") == PROFILE_VERSION
            else ""
        )
        if not self.device_fp:
            self._save()

    async def ensure_fingerprint(self, client: httpx.AsyncClient) -> str:
        if self.device_fp:
            return self.device_fp
        response = await client.post(DEVICE_FP_URL, json=self._payload())
        response.raise_for_status()
        payload: dict[str, Any] = response.json()
        data = payload.get("data") or {}
        fingerprint = str(data.get("device_fp") or "")
        if payload.get("retcode") != 0 or not fingerprint:
            raise AppError("device_fp_failed", "米游社设备注册失败，请稍后重试", 502)
        self.device_fp = fingerprint
        self._save()
        return fingerprint

    def _payload(self) -> dict[str, str]:
        return {
            "device_id": self.fp_device_id,
            "seed_id": str(uuid.uuid4()),
            "seed_time": str(int(time.time() * 1000)),
            "platform": "2",
            "device_fp": secrets.token_hex(7)[:13],
            "app_name": "bbs_cn",
            "bbs_device_id": self.device_id,
            "ext_fields": json.dumps(self._android_fields(), separators=(",", ":")),
        }

    def _android_fields(self) -> dict[str, str | int]:
        return {
            "proxyStatus": 0,
            "isRoot": 0,
            "romCapacity": "512",
            "deviceName": self.device_name,
            "productName": self.product_name,
            "romRemain": "459",
            "manufacturer": "XiaoMi",
            "appMemory": "512",
            "hostname": "android-build",
            "screenSize": "1440x2905",
            "osVersion": "14",
            "aaid": "",
            "vendor": "unknown",
            "accelerometer": "1.48x7.17x6.28",
            "buildTags": "release-keys",
            "packageName": "com.mihoyo.hyperion",
            "networkType": "WiFi",
            "model": self.device_name,
            "brand": "XiaoMi",
            "oaid": "",
            "hardware": "qcom",
            "deviceType": "OP5913L1",
            "devId": "REL",
            "serialNumber": "unknown",
            "buildTime": "1693626947000",
            "buildUser": "android-build",
            "ramCapacity": "469679",
            "magnetometer": "20.08x-27.48x2.19",
            "display": f"{self.product_name}_14_release-keys",
            "ramRemain": "239814",
            "deviceInfo": (
                f"XiaoMi/{self.product_name}/OP5913L1:14/"
                "SKQ1.221119.001/release-keys"
            ),
            "gyroscope": "0.03x0.01x0.01",
            "vaid": "",
            "buildType": "user",
            "sdkVersion": "34",
            "board": "taro",
        }

    @staticmethod
    def _random_text(length: int) -> str:
        alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return "".join(secrets.choice(alphabet) for _ in range(length))

    def _load(self) -> dict[str, Any]:
        try:
            value = json.loads(self.path.read_text(encoding="utf-8"))
            return value if isinstance(value, dict) else {}
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            return {}

    def _save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        temporary = self.path.with_suffix(".tmp")
        temporary.write_text(
            json.dumps(
                {
                    "profile": PROFILE_VERSION,
                    "device_id": self.device_id,
                    "fp_device_id": self.fp_device_id,
                    "device_name": self.device_name,
                    "product_name": self.product_name,
                    "device_fp": self.device_fp,
                },
                ensure_ascii=False,
            ),
            encoding="utf-8",
        )
        os.chmod(temporary, 0o600)
        temporary.replace(self.path)

File: backend/src/providers/device.ts
import { chmodSync, existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { randomBytes, randomUUID } from "node:crypto";
import { AppError } from "../core/errors";

interface Saved { profile?: string; device_id?: string; fp_device_id?: string; hoyoplay_device_id?: string; device_name?: string; product_name?: string; device_fp?: string }
const text = (length: number): string => randomBytes(length).toString("base64url").toUpperCase().slice(0, length);
const lowerText = (length: number): string => [...randomBytes(length)].map((value) => "0123456789abcdefghijklmnopqrstuvwxyz"[value % 36]).join("");

export class Device {
  readonly deviceId: string; readonly fpDeviceId: string; readonly hoyoplayDeviceId: string; readonly deviceName: string; readonly productName: string;
  deviceFP: string;
  constructor(private readonly path: string) {
    const value = this.load(); this.deviceId = value.device_id ?? randomUUID(); this.fpDeviceId = value.fp_device_id ?? randomBytes(8).toString("hex");
    this.hoyoplayDeviceId = value.hoyoplay_device_id?.match(/^[0-9a-z]{53}$/) ? value.hoyoplay_device_id : lowerText(53);
    this.deviceName = value.device_name ?? text(12); this.productName = value.product_name ?? text(6);
    this.deviceFP = value.profile === "snap-hutao-android-v2" ? value.device_fp ?? "" : "";
    if (!value.hoyoplay_device_id || !this.deviceFP) this.save();
  }

  get loginDeviceId(): string {
    return this.hoyoplayDeviceId;
  }

  async fingerprint(): Promise<string> {
    if (this.deviceFP) return this.deviceFP;
    const response = await fetch("https://public-data-api.mihoyo.com/device-fp/api/getFp", { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(this.payload()) });
    const body = await response.json() as { retcode?: number; data?: { device_fp?: string } };
    if (!response.ok || body.retcode !== 0 || !body.data?.device_fp) throw new AppError("device_fp_failed", "米游社设备注册失败，请稍后重试", 502);
    this.deviceFP = body.data.device_fp; this.save(); return this.deviceFP;
  }

  private payload(): Record<string, string> {
    const ext = { proxyStatus: 0, isRoot: 0, romCapacity: "512", deviceName: this.deviceName, productName: this.productName,
      romRemain: "512", hostname: "dg02-pool03-kvm87", screenSize: "1440x2905", isTablet: 0, aaid: "", model: this.deviceName,
      brand: "XiaoMi", hardware: "qcom", deviceType: "OP5913L1", devId: "REL", serialNumber: "unknown", sdCapacity: 512215,
      buildTime: "1693626947000", buildUser: "android-build", simState: 5, ramRemain: "239814", appUpdateTimeDiff: 1702604034482,
      deviceInfo: `XiaoMi/${this.productName}/OP5913L1:13/SKQ1.221119.001/T.118e6c7-5aa23-73911:user/release-keys`,
      vaid: "", buildType: "user", sdkVersion: "34", ui_mode: "UI_MODE_TYPE_NORMAL", isMockLocation: 0, cpuType: "arm64-v8a",
      isAirMode: 0, ringMode: 2, chargeStatus: 1, manufacturer: "XiaoMi", emulatorStatus: 0, appMemory: "512", osVersion: "14",
      vendor: "unknown", accelerometer: "1.4883357x7.1712894x6.2847486", sdRemain: 239600, buildTags: "release-keys",
      packageName: "com.mihoyo.hyperion", networkType: "WiFi", oaid: "", debugStatus: 1, ramCapacity: "469679",
      magnetometer: "20.081251x-27.487501x2.1937501", display: `${this.productName}_13.1.0.181(CN01)`,
      appInstallTimeDiff: 1688455751496, packageVersion: "2.20.1", gyroscope: "0.030226856x0.014647375x0.010652636",
      batteryStatus: 100, hasKeyboard: 0, board: "taro" };
    return { device_id: this.fpDeviceId, seed_id: randomUUID(), seed_time: String(Date.now()), platform: "2", device_fp: randomBytes(7).toString("hex").slice(0, 13),
      app_name: "bbs_cn", bbs_device_id: this.deviceId, ext_fields: JSON.stringify(ext) };
  }

  private load(): Saved { try { return existsSync(this.path) ? JSON.parse(readFileSync(this.path, "utf8")) as Saved : {}; } catch { return {}; } }
  private save(): void {
    mkdirSync(dirname(this.path), { recursive: true }); const temporary = `${this.path}.tmp`;
    writeFileSync(temporary, JSON.stringify({ profile: "snap-hutao-android-v2", device_id: this.deviceId, fp_device_id: this.fpDeviceId, hoyoplay_device_id: this.hoyoplayDeviceId,
      device_name: this.deviceName, product_name: this.productName, device_fp: this.deviceFP }));
    chmodSync(temporary, 0o600); renameSync(temporary, this.path);
  }
}
